File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        auto curr = head;
        int count = 0;
        while (curr) {
            ++count;
            curr = curr->next;
        }
        count -= n;
        auto dummy = new ListNode(0, head);
        curr = dummy;
        for (int i = 0; i < count; ++i) {
            curr = curr->next;
        }
        auto toDelete = curr->next;
        curr->next = toDelete->next;
        delete toDelete;
        auto toReturn = dummy->next;
        delete dummy;
        return toReturn;
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-3.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        curr = head
        dummy = Node(0)
        newCurr = dummy
        m = {}
        while curr:
            newCurr.next = Node(curr.val)
            newCurr = newCurr.next
            m[curr] = newCurr
            curr = curr.next
        curr = head
        newCurr = dummy.next
        while curr:
            newCurr.random = None if not curr.random else m[curr.random]
            curr = curr.next
            newCurr = newCurr.next
        return dummy.next